drivers/pwm/pwm-ts.c

```diff
@@ -90,49 +83,36 @@ static int ts_pwm_calc(struct ts_pwm *ts,
 
 		dev_dbg(ts->chip.dev, "shift=%u cnt=%u duty_cnt=%u\n",
 			shift, cnt, duty_cnt);
-
-		ts->shift = shift;
-		ts->period = cnt;
-		ts->duty = cnt - duty_cnt;
+		duty_reg = cnt - duty_cnt;
 	}
 
+	writew(cnt, ts->base + REG_PERIOD);
+	writew(duty_reg, ts->base + REG_DUTY);
+	writew(shift, ts->base + REG_SHIFT);
```

Comment:
I'm curious what this looks like on the PWM line when there is a change? In looking at other PWM drivers, it looks like in general they rely on hardware to only apply settings at a specific time. I'm not sure what our core does in this case. If it is really glitchy, we might consider disabling the PWM, applying settings, and then enabling the PWM?

Reply:
The PWM framework enforces that the period does not change when the PWM channel is enabled. I based this on pwm-mxs which does seem to write its own registers similarly. The period changing is all that can influence our period/shift registers having different values, so the only piece that can change is duty. The FPGA core is safe to change duty at any time. I could update this in code to check if its enabled, but I'd have to store the cnt/cycle values from when period actually is set. Its certainly doable to store & have different paths for if its enabled or not to save the extra calculation / write when the user is only changing duty, but I'm tempted to leave that until we have a need to change duty faster.

Reply:
No, no, that should be okay.
> The PWM framework enforces that the period does not change when the PWM channel is enabled.

I was unaware of that, and as shift is only affected by period, then only duty would ever change and all of this should be safe.